### PR TITLE
fix(codegen): Support `extern crate serde` not in toplevel module

### PR DESCRIPTION
**Pull request by dtolnay**
Addresses #159.

The generated code looks like:

```rust
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _IMPL_DESERIALIZE_FOR_MyStruct: () = {
    extern crate serde as _serde;
    #[automatically_derived]
    impl _serde::de::Deserialize for MyStruct {
        // ...
    }
};
```

### CODE REVIEW

File: serde_tests/tests/test.rs [hunk 1]
```diff
@@ -1,6 +1,4 @@
 #![cfg_attr(feature = "nightly", feature(plugin))]
 #![cfg_attr(feature = "nightly", plugin(clippy))]
 
-extern crate serde;
```
Comment (oli-obk) [inline serde_tests/tests/test.rs]: why was this removed? Did it cause an error?

**Comment by oli-obk** [comment]
Does this break any existing code? You changed quite a few tests' imports.

File: serde_tests/tests/test.rs (hunk 1 above)
Comment (dtolnay) [inline serde_tests/tests/test.rs]: No error, just to demonstrate that things work even without "extern crate serde" at the top level. I kept the equivalent line in [bench.rs](https://github.com/dtolnay/serde/blob/5eb1d7dd4b9d5aa19599be10771fb020a39c46a3/serde_tests/benches/bench.rs#L6) so the build will still uses both approaches.

**Comment by dtolnay** [comment]
This does not break any existing code.

**Comment by dtolnay** [comment]
This makes `extern crate serde` optional unless you are using it explicitly in your own code, so I have removed it from the example in the readme.

**Comment by erickt** [comment]
@dtolnay: Once again, great work here. I'm working on merging this patch in by hand.

**Comment by dtolnay** [comment]
@erickt I fixed the merge conflict myself.

**Comment by dtolnay** [comment]
Ping @erickt - looks like you said this was ready to merge. Once you merge it, I will resolve the merge conflict in https://github.com/serde-rs/serde/pull/293#issuecomment-214442429.

**Comment by erickt** [comment]
Thanks @dtolnay! I've been swamped with work recently, and unfortunately that's resulted in lots of my projects suffering. Thanks for keeping up with this.
